sdk/identity/azure-identity/src/azure_cli_credential.cpp

```diff
@@ -0,0 +1,654 @@
+std::string RunShellCommand(
+    std::string const& command,
+    DateTime::duration timeout,
+    Context const& context);
+}
+
+AccessToken AzureCliCredential::GetToken(
+    TokenRequestContext const& tokenRequestContext,
+    Context const& context) const
+{
+  try
+  {
+    auto const azCliResult = RunShellCommand(
+        GetAzCommand(
+            TokenCredentialImpl::FormatScopes(tokenRequestContext.Scopes, true, false), m_tenantId),
+        m_cliProcessTimeout,
+        context);
+
+    try
+    {
+      return TokenCredentialImpl::ParseToken(azCliResult, "accessToken", "expiresIn", "expiresOn");
+    }
+    catch (std::exception const&)
+    {
+      // Throw the az command output (error message)
+      // limited to 250 characters (250 has no special meaning).
+      throw std::runtime_error(azCliResult.substr(0, 250));
+    }
+  }
+  catch (std::exception const& e)
+  {
+    throw AuthenticationException(std::string("AzureCliCredential::GetToken(): ") + e.what());
+  }
+}
+
+namespace {
+class ShellProcess;
+class OutputPipe final {
+  friend class ShellProcess;
+
+private:
+#if defined(AZ_PLATFORM_WINDOWS)
+  HANDLE m_writeHandle = NULL;
+  HANDLE m_readHandle = NULL;
+  OVERLAPPED m_overlapped = {};
+#else
+  std::vector<int> m_fd;
+#endif
+
+  OutputPipe(OutputPipe const&) = delete;
+  OutputPipe& operator=(OutputPipe const&) = delete;
+
+public:
+  OutputPipe();
+
+  ~OutputPipe();
+
+  bool NonBlockingRead(
+      std::vector<std::string::value_type>& buffer,
+      std::string::size_type& bytesRead,
+      bool& willHaveMoreData);
+};
+
+class ShellProcess final {
+private:
+#if defined(AZ_PLATFORM_WINDOWS)
+  HANDLE m_processHandle = NULL;
+#else
+  std::vector<char*> m_argv;
+  std::vector<char> m_argvValues;
+
+  std::vector<char*> m_envp;
+  std::vector<char> m_envpValues;
+
+  posix_spawn_file_actions_t m_actions = {};
+  int m_pid = -1;
+#endif
+
+  ShellProcess(ShellProcess const&) = delete;
+  ShellProcess& operator=(ShellProcess const&) = delete;
+
+public:
+  ShellProcess(std::string const& command, OutputPipe& outputPipe);
+  ~ShellProcess();
+
+  void Terminate();
+};
+
+std::string RunShellCommand(
+    std::string const& command,
+    DateTime::duration timeout,
+    Context const& context)
+{
+  // Use steady_clock so we're not affected by system time rewinding.
+  auto const terminateAfter = std::chrono::steady_clock::now()
+      + std::chrono::duration_cast<std::chrono::steady_clock::duration>(timeout);
+
+  std::string output;
+
+  OutputPipe pipe;
+  ShellProcess shellProcess(command, pipe);
+
+  // Typically token json is just a bit less than 2KiB.
+  // The best buffer size is the one that lets us to read it in one go.
+  // (Should it be smaller, we will succeed as well, it'll just take more iterations).
+  std::vector<std::string::value_type> processOutputBuf(2 * 1024);
+
+  auto willHaveMoreData = true;
+  do
+  {
+    // Check if we should terminate
+    {
+      if (context.IsCancelled())
```

Comment:
Why have two separate cancellation states? Why not trust the caller's cancellation state and if the provided context doesn't have a timeout, provide your own?

Reply:
Context comes from the client method. Timeout comes from the credential options. Other SDKs have it this way too.